### PR TITLE
Space - Lak

### PR DESCRIPTION
**Pull request by easternwashingaden**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I created some custom model methods to self.top_ten_list(category), self.spot_light, self.order_by_votes_count(category), and self.count_votes(work_id).
Describe how you approached testing that model method. What edge cases did you come up with? | I approached the testing that model methods by setting some data in the fixture files for users, works, and votes. From the fixture, I know and expect what work has the most votes, what the top ten are, how many vote counts each work has based on its category, and I wrote my tests based on that information. One example of my edge case is i expect my vote count function to return 0 if there is no vote associated to that work. 
What are `session` and `flash`? What is the difference between them? | session is a special type of hash-like-object that we utilize to keep track of data throughout a users "session", which normally ends when they close their browser. session is similar to flash, except that data stored there will not go away after the next request-response cycle. Flash is available the rest of this request cycle and the next complete request cycle where session is available until the user closes the browser.
What was one thing that you gained more clarity on through this assignment? | using flash, error message, nested routes, custom model tests, fixture, session, and more...
What is the Heroku URL of your deployed application? | https://lak-media-ranker.herokuapp.com

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: app/views/works/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+          </tr>
+        <% end %>
+      </tbody>
+    </table>
+  <% else %>
+    <p>No work in the system!</p>
+  <% end %>
+
+  <h4>Books</h4>
+  <% if @works.where(category: "book").length > 0 %>
```
Comment (beccaelenzil) [inline app/views/works/index.html.erb:32]: Note that you have nearly the same code repeated 3 times. Consider how you could use a loop or a partial view to DRY this up.

File: app/controllers/votes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
+class VotesController < ApplicationController
+  def new
+    @vote = Vote.new
+  end
+
+  def create
+    # Check if there is a user logged in
+    if session[:user_id]
+      @user = User.find_by(id: session[:user_id])
```
Comment (beccaelenzil) [inline app/controllers/votes_controller.rb:9]: You need to find the current user in multiple controllers. This would be great functionality to put in a controller filter in the `application_controller`.

File: app/models/work.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  has_many :votes
+  has_many :users, through: :votes
+
+  validates :category, presence: true
+  validates :title, presence: true, uniqueness: { scope: :category} # Allow user to add the same existing title with different category
+  validates :creator, presence: true
+  validates :publication_year, presence: true
+  validates :description, presence: true
+
+  def self.top_ten_list(category)
```
Comment (beccaelenzil) [inline app/models/work.rb:11]: These are a great set of custom methods. Good choice making category an argument to DRY up your code.

File: test/models/work_test.rb [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+      it "returns the top ten list of albums that have most votes in descending order " do
+        # movie has 0 votes
+        top_movie= works(:movie1)
+        expect(Work.top_ten_list(:movie).first).must_equal top_movie
+        expect(Work.top_ten_list(:movie).first.votes.count).must_equal 0
+      end
+
+    end
+    describe "spot_light" do
+      it "returns one Work that has the most votes" do
```
Comment (beccaelenzil) [inline test/models/work_test.rb:191]: Consider a test for if their are tied works.

File: test/models/user_test.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    user = User.first
+    [:name].each do |field|
+
+      # Assert
+      expect(user).must_respond_to field
+    end
+  end
+
+  describe "relationships" do
+    it "can have many votes and have many works through votes" do
```
Comment (beccaelenzil) [inline test/models/user_test.rb:25]: Consider splitting up works and votes into two separate tests since they're two separate relationships.

**Comment by beccaelenzil** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️Consider redirecting to the show page when you create a new work.
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️redirect
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | See in-line comment
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | See in-line comment
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Logical/Organized | ✅

**Review by beccaelenzil — COMMENTED** [review]
Great work on this project. Your code is well written, well tested, and closely matches the example site. You've written great custom model methods. I've left a few in-line comments focused on how you could use controller filters and view partials to DRY up your code. Please let me know if you have any questions. Keep up the hard work!
